benchmark: Condtionalize AMD64 Time Stamp Counter

peterd-sky opened the pull request:
Use a Time Stamp Counter only on AMD64 based platforms. Other platforms
will have the ability to run the benchmark, but zeros will be reported
as time values. A warning is printed at compile time to notify the user
that Time Stamp Counters are not available for non-AMD64 based platforms.

peterd-sky committed:
benchmark: Condtionalize AMD64 Time Stamp Counter

Use a Time Stamp Counter only on AMD64 based platforms. Other platforms
will have the ability to run the benchmark, but zeros will be reported
as time values. A warning is printed at compile time to notify the user
that Time Stamp Counters are not available for non-AMD64 based platforms.

msune commented on test/benchmark.cc:
```diff
@@ -53,6 +53,8 @@ CDADA_MAP_CUSTOM_TYPE_DECL(test_u3552_t);
 CDADA_SET_CUSTOM_TYPE_DECL(test_u3552_t);
 CDADA_STACK_CUSTOM_TYPE_DECL(test_u3552_t);
 
+#ifdef __amd64__
```
I think this is also valid for x86

peterd-sky replied:
There are many preprocessor macros defined for both `amd` and `x86`. See https://sourceforge.net/p/predef/wiki/Architectures/. It seemed a bit of overkill to add them all. What do you think?

msune replied:
Certainly the predefined variabless preprocessor path seems to convoluted.

It might be cleaner to run an `AC_COMPILE_IFELSE` in `configure.ac` with a very simple program trying to use `__rdtsc`, and then create a MACRO as a result. If that would be in a separate m4, e.g. `config_m4/rdtsc.m4`, it would be ideal

peterd-sky replied:
Done. Please see code diffs.

msune reviewed (commented):
Hi Peter,

Thank you for the contribution. I've changed the target branch to `devel` instead.

I only have a question wrt to the `#ifdef` itself for x86. The approach seems fine.

Question, is the rest of `libcdada` compiling fine in the non-amd64 arch you are compiling? Which is it?

Unfortunately, travis has some limitations on which archs tests can be run - hence my interest.

peterd-sky commented:
Hi Marc,

The rest of `libcdada` is compiling without warning or error for the PPC using the enea cross tools for the xxx and yyy.

peterd-sky committed:
rdtsc: Add config_m4 script to autodetect RDTSC

Add a config_m4 script that autodetects if RDTSC support is available.
If it is, the HAVE_RDTSC preprocessor macro is defined. If it is not, the
preprocessor macro is not defined.

msune commented:
@peterd-sky I tried your last patch. Overall looks great, just that with an i686 container, the following warnings are seen in config:

```
configure:13291: checking whether RDTSC support is available
configure:13308: gcc -o conftest -std=gnu89 -Werror -Wall -g -O2 -O3   conftest.c  >&5
conftest.c: In function 'main':
conftest.c:36:14: error: format '%lu' expects argument of type 'long unsigned int', but argument 2 has type 'unsigned int' [-Werror=format=]
    printf("%lu, %u\n", hi, lo);
            ~~^         ~~
            %u
cc1: all warnings being treated as errors
```

I suspect it's `printf("%u, %u", hi, lo)`

msune commented:
Missed that; I tried it ` docker run -it i386/debian`

peterd-sky commented:
> printf("%lu, %u\n", hi, lo);

I had changed to `%lu` to cause a failure and obviously I forgot to remove it. I did this to simulate a test on platforms that do not support the rdtsc register.

peterd-sky committed:
config: Fix error in rdtsc.m4 script

Remove forced error from the rdtsc.m4 script so that the RDTSC support
is properly detected in amd64 and x86 based platforms.

msune commented:
ACK
